src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleDictionary.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonSerializer
+    {
+        private static void HandleStartDictionary(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack state)
+        {
+            Debug.Assert(!state.Current.IsProcessingEnumerable);
+
+            JsonPropertyInfo jsonPropertyInfo = state.Current.JsonPropertyInfo;
+            if (jsonPropertyInfo == null)
+            {
+                jsonPropertyInfo = state.Current.JsonClassInfo.CreateRootObject(options);
+            }
+
+            Debug.Assert(jsonPropertyInfo != null);
+
+            // A nested object or dictionary so push new frame.
+            if (state.Current.PropertyInitialized)
+            {
+                Debug.Assert(state.Current.IsDictionary);
+
+                JsonClassInfo classInfoTemp = state.Current.JsonClassInfo;
+                state.Push();
+                state.Current.JsonClassInfo = classInfoTemp.ElementClassInfo;
+                state.Current.InitializeJsonPropertyInfo();
+
+                ClassType classType = state.Current.JsonClassInfo.ClassType;
+                if (classType == ClassType.Value &&
+                    jsonPropertyInfo.ElementClassInfo.Type != typeof(object) &&
+                    jsonPropertyInfo.ElementClassInfo.Type != typeof(JsonElement))
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.JsonClassInfo.Type, reader, state.PropertyPath);
+                }
+
+                JsonClassInfo classInfo = state.Current.JsonClassInfo;
+                state.Current.ReturnValue = classInfo.CreateObject();
+                return;
+            }
+
+            state.Current.PropertyInitialized = true;
+
+            // If current property is already set (from a constructor, for example) leave as-is.
+            if (jsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue) == null)
```

Comment:
Store `state.Current.ReturnValue` in a local?

Reply:
What happens if `state.Current.ReturnValue == null` here? Couldn't a cast like this null ref:
https://github.com/dotnet/corefx/blob/0d3caeba10c3c4a83f16a11c4298e39528d923b5/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs#L76

Reply:
Storing `state.Current.ReturnValue` in a local I assume you mean for performance?

For the null question, doing a `((TClass)null)` works fine -- just a null object reference.